crates/interpreter/src/valid.rs

```diff
@@ -30,6 +30,57 @@ pub fn validate(binary: &[u8]) -> Result<(), Error> {
 type Parser<'m> = parser::Parser<'m, Check>;
 type CheckResult = MResult<(), Check>;
 
+#[allow(dead_code)]
+#[derive(Default, Copy, Clone)]
+pub struct SideTableEntry {
+    data: u32,
+}
+
+#[allow(dead_code)]
+impl SideTableEntry {
+    const DELTA_IP_MASK: u32 = 0xFFFF; // 16 bits for delta_ip
+    const DELTA_STP_MASK: u32 = 0x3F << 16; // 6 bits for delta_stp
+    const VAL_CNT_MASK: u32 = 0x1F << 22; // 5 bits for val_cnt
+    const POP_CNT_MASK: u32 = 0x1F << 27; // 5 bits for pop_cnt
+
+    #[allow(clippy::manual_range_contains)]
+    fn new(delta_ip: i32, delta_stp: i32, val_cnt: u32, pop_cnt: u32) -> Self {
+        assert!(delta_ip >= -32768 && delta_ip <= 32767);
+        assert!(delta_stp >= -32 && delta_stp <= 31);
+        assert!(val_cnt <= 31);
+        assert!(pop_cnt <= 31);
+
+        let delta_ip = (delta_ip as u32) & Self::DELTA_IP_MASK;
+        let delta_stp = ((delta_stp as u32) & 0x3F) << 16;
+        let val_cnt = (val_cnt & 0x1F) << 22;
+        let pop_cnt = (pop_cnt & 0x1F) << 27;
+
+        SideTableEntry { data: delta_ip | delta_stp | val_cnt | pop_cnt }
+    }
+
+    /// Get the amount to adjust the instruction pointer by if the branch is taken.
+    fn get_delta_ip(&self) -> i32 {
+        (self.data & Self::DELTA_IP_MASK) as i32
+            - (((self.data & Self::DELTA_IP_MASK) & 0x8000) << 16) as i32
+    }
+
+    /// Get the amount to adjust the side-table pointer by if the branch is taken.
+    fn get_delta_stp(&self) -> i32 {
+        ((self.data & Self::DELTA_STP_MASK) >> 16) as i32
+            - ((((self.data & Self::DELTA_STP_MASK) >> 16) & 0x20) << 6) as i32
+    }
+
+    /// Get the number of values that will be copied if the branch is taken.
+    fn get_val_cnt(&self) -> u32 {
+        (self.data & Self::VAL_CNT_MASK) >> 22
+    }
+
+    /// Get the number of values that will be popped if the branch is taken.
+    fn get_pop_cnt(&self) -> u32 {
+        (self.data & Self::POP_CNT_MASK) >> 27
+    }
+}
```

Comment:
I believe this logic could be simplified by storing only positive values and avoiding sign-extension.

```rust
pub struct SideTableEntry(u32);
pub struct SideTableEntryView {
    pub delta_ip: i32,
    pub delta_stp_mask: i32,
    pub val_cnt: u32,
    pub pop_cnt: u32,
}

impl SideTableEntry {
    const DELTA_IP_MASK: u32 = 0x0000ffff;
    const DELTA_STP_MASK: u32 = 0x003f0000;
    const VAL_CNT_MASK: u32 = 0x07c00000;
    const POP_CNT_MASK: u32 = 0xf8000000;

    fn new(view: SideTableEntryView) -> Result<Self, Result> {
        let mut fields = 0;
        fields |= Self::into_signed_field(Self::DELTA_IP_MASK, view.delta_ip)?;
        fields |= Self::into_signed_field(Self::DELTA_STP_MASK, view.delta_stp)?;
        fields |= Self::into_field(Self::VAL_CNT_MASK, view.val_cnt)?;
        fields |= Self::into_field(Self::POP_CNT_MASK, view.pop_cnt)?;
        SideTableEntry(fields)
    }

    fn view(self) -> SideTableEntryView {
        let delta_ip = Self::from_signed_field(Self::DELTA_IP_MASK, self.0);
        // ...
        SideTableEntryView { delta_ip, delta_stp, val_cnt, pop_cnt }
    }

    fn into_signed_field(mask: u32, value: i32) -> Result<u32, Error> {
        let value = value.wrapping_add(Self::offset(mask));
        Self::into_field(mask, value)
    }

    fn from_signed_field(mask: u32, field: u32) -> i32 {
        Self::from_field(mask, field) as i32 - Self::offset(mask)
    }

    fn offset(mask: u32) -> i32 {
        1 << (mask.count_ones() - 1)
    }

    fn into_field(mask: u32, value: u32) -> Result<u32, Error> {
        let field = (value << mask.trailing_zeros()) & mask;
        if from_field(mask, field) != value {
            return Err(unsupported(if_debug!(Unsupported::SideTable)));
        }
        Ok(field)
    }

    fn from_field(mask: u32, field: u32) -> u32 {
        (field & mask) >> mask.trailing_zeros()
    }
}
```